syntax = "proto2";

message CityEmpty {}

message CityIdRequest {
  required string id = 1;
}

message CityNotIdRequest {
  required string state = 1;
  required string name = 2;
}


message CityRequest {
  required string id = 1;
  required string state = 2;
  required string name = 3;
}


message CityTableRequest {
  required int32 page = 1;
  optional int32 per_page = 2 [default = 15];
  optional string search = 3;
}

message CityResponse {
  required CityRequest city = 1;
}

message CityMultipleResponse {
  repeated CityRequest city = 1;
}

message CityTableResponse {
  repeated CityRequest items = 1;
  required int32 page = 2;
  required int32 per_page = 3;
  required int32 total_items = 4;
  required int32 num_pages = 5;
}

service City {
  rpc table (CityTableRequest) returns (CityTableResponse);
  rpc get_all (CityEmpty) returns (CityMultipleResponse);
  rpc get (CityIdRequest) returns (CityResponse);
  rpc save (CityNotIdRequest) returns (CityResponse);
  rpc update (CityRequest) returns (CityResponse);
  rpc delete (CityIdRequest) returns (CityEmpty);
}
